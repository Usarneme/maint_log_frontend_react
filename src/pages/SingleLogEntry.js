import React from 'react'
import { Link, Redirect, useParams } from 'react-router-dom'
import moment from 'moment'

import PhotoEditor from '../components/PhotoEditor'
import VehicleHeader from '../components/VehicleHeader'
import EditPencil from '../images/editPencil.svg'

import '../styles/singleLogEntry.css'

function SingleLogEntry(props) {
  const { slug } = useParams()

  const isLoggedIn = (props.user && props.user.cookies ? props.user.cookies.length > 0 : false)
  if (!isLoggedIn) return <Redirect to="/welcome" />

  console.log(`Displaying log entry: ${slug}`)
  const log = props.user.log.filter(entry => entry.slug === slug)
  const { id, shortDescription, longDescription, dateStarted, dateCompleted, dateEntered, dateDue, mileageDue, name, odometer, tools, parts, partsCost, laborCost, serviceLocation, photos } = log[0]
  let vehicle = {}
  if (props.user.vehicle && props.user.vehicle[0]) {
    vehicle = props.user.vehicle[0]
  }

  return (
    <div className="inner">
      <h2>{name.length > 120 ? `${name.substring(0,120)}...` : name}</h2>
      <VehicleHeader vehicle={vehicle} />
      <div className="single__details">
        <p>
          <strong>Short Description: </strong>
          <span>{shortDescription}</span>
        </p>
        <p>
          <strong>Long Description: </strong>
          <span>{longDescription}</span>
        </p>
        <p>
          <strong>Date Started: </strong>
          <span>{moment(dateStarted).format("MMM Do YYYY")}</span>
        </p>
        <p>
          <strong>Date Completed: </strong>
          <span>{moment(dateCompleted).format("MMM Do YYYY")}</span>
        </p>
        <p>
          <strong>Date Entered into Log: </strong>
          <span>{moment(dateEntered).format("MMM Do YYYY")}</span>
        </p>
        <p>
          <strong>Date When Service is Due Next: </strong>
          <span>{moment(dateDue).format("MMM Do YYYY")}</span>
        </p>
        <p>
          <strong>Mileage When Service is Due Next: </strong>
          <span>{Number(mileageDue).toLocaleString()}</span>
        </p>
        <p>
          <strong>Tools: </strong>
          <span>{tools}</span>
        </p>
        <p>
          <strong>Parts: </strong>
          <span>{parts}</span>
        </p>
        <p>
          <strong>Parts Cost: </strong>
          <span>${partsCost}</span>
        </p>
        <p>
          <strong>Labor Cost: </strong>
          <span>${laborCost}</span>
        </p>
        <p>
          <strong>Service Location: </strong>
          <span>{serviceLocation}</span>
        </p>
        <p>
          <strong>Odometer: </strong>
          <span>{Number(odometer).toLocaleString()}</span>
        </p>
        <Link className="button editPencil" to={`/log/${id}/edit`}>
          <img src={EditPencil} alt="Edit Pencil" className="svg" />
          <span>Edit</span>
        </Link>

        { photos && <PhotoEditor photos={photos} editingBlocked={true} /> }
        <input type="hidden" name="previousPhotos" value={photos.toString()} />
      </div>

    </div>
  )
}

export default SingleLogEntry 