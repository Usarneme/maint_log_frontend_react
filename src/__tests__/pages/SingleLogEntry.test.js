import React from 'react'
import TestRenderer from 'react-test-renderer'
import { BrowserRouter as Router } from 'react-router-dom'

import SingleLogEntry from '../../pages/SingleLogEntry'

describe('SINGLE LOG ENTRY PAGE', () => {
  const mockMatch = {
    "path":"/log/:id/edit",
    "url":"/log/abc123/edit",
    "isExact":true,
    "params":{"id":"abc123","slug":"abc123"}
  }
  const mockLocation = {
    pathname: '/log/abc123/edit',
    hash: '',
    search: '',
    state: '',
    params: {
      'id': 'abc123',
      'slug': 'abc123'
    }
  }
  const props = {
    "match": mockMatch,
    "location": mockLocation,
    "history": {},
    "user": {
      "log": [
        {
          "tools":["Socket wrench, 1/2\" socket, screwdriver, WD40, Tri-flow, brakleen, shop towels, gloves"],
          "parts":["None"],
          "receipts":[""],
          "photos":["1d124d0b-6495-4c17-8145-1d2d5dbbcbde.jpeg","96b45924-a322-4560-a8df-4a2a6dc016fa.jpeg","e0e09d60-16e5-4f2a-b60a-c8eeff05eed7.jpeg"],
          "_id":"abc123",
          "name":"Door latches refurb",
          "vehicle":"5e4cee3fd2ebbc8f5e62f214",
          "dateStarted":"2020-03-11T00:00:00.000Z",
          "dateCompleted":"2020-04-01T00:00:00.000Z",
          "dateDue":null,
          "mileageDue":null,
          "shortDescription":"Removed, degreased, cleaned, lubricated, and re-installed the door latches.",
          "longDescription":"Reinstall.",
          "partsCost":0,
          "laborCost":0,
          "serviceLocation":"Self",
          "odometer":219994,
          "author":"5e3ce3f10ac0bc29266d5287",
          "dateEntered":"2020-04-02T00:00:00.000Z",
          "slug":"abc123",
          "__v":0,
          "id":"abc123"
        },
        {
          "tools":["Socket"],
          "parts":["None"],
          "receipts":[""],
          "photos":["1d124d0b"],
          "_id":"xyz789",
          "name":"Door",
          "vehicle":"5e4cee3fd2ebbc8f5e62f214",
          "dateStarted":"2020",
          "dateCompleted":"2020",
          "dateDue":null,
          "mileageDue":null,
          "shortDescription":"Removed",
          "longDescription":"Reinstall",
          "partsCost":0,
          "laborCost":0,
          "serviceLocation":"Self",
          "odometer":219994,
          "author":"5e3ce3f10ac0bc29266d5287",
          "dateEntered":"2020",
          "slug":"test-two",
          "__v":0,
          "id":"xyz789"
        }
      ]
    },
    "updateUserState": () => {}
  }

  it('** Renders the Edit Page', () => {
    const raw = TestRenderer.create(<Router><SingleLogEntry {...props} /></Router>)
    expect(raw.toTree().rendered.rendered.type.name).toBe('SingleLogEntry')
  })

  it('** Gets the log slug from the url params', () => {
    const raw = TestRenderer.create(<Router><SingleLogEntry {...props} /></Router>)
    expect(raw.toTree().rendered.rendered.props).toStrictEqual(props)
    expect(raw.toTree().rendered.rendered.rendered.props.children[0].props.children).toBe(props.user.log[0].name)
  })
})