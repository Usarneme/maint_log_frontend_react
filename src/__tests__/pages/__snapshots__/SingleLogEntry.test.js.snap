// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SINGLE LOG ENTRY PAGE 	Not Logged In redirects to / route 1`] = `
Array [
  <a
    className="siteTitle"
    href="/"
    onClick={[Function]}
  >
    <div
      aria-label="Home Icon"
      className="svg"
      role="img"
    />
    <h1>
      Vehicle Maintenance Log
    </h1>
  </a>,
  <div
    className="inner"
  >
    <div
      className="card guest__options"
    >
      <div
        className="button guest__option login guest__option__active"
        onClick={[Function]}
      >
        <h4>
          Login & View Your Maintenance Log
        </h4>
        <div
          aria-label="Login Icon"
          role="img"
        />
      </div>
      <div
        className="button guest__option register "
        onClick={[Function]}
      >
        <h4>
          Register & Start Your Maintenance Log
        </h4>
        <div
          aria-label="Register Icon"
          role="img"
        />
      </div>
    </div>
    <div
      className=""
    >
      <form
        className="card"
        method="POST"
        onSubmit={[Function]}
      >
        <h2>
          Login
        </h2>
        <label
          htmlFor="email"
        >
          Email Address
        </label>
        <input
          name="email"
          onChange={[Function]}
          placeholder="Enter email..."
          type="email"
          value=""
        />
        <label
          htmlFor="password"
        >
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Enter password..."
          type="password"
          value=""
        />
        <input
          className="button"
          type="submit"
          value="Log In â†’"
        />
        <button
          className="button forgot__password__button"
          onClick={[Function]}
        >
          Forgot Your Password?
        </button>
      </form>
    </div>
  </div>,
]
`;
